Generic Vulkan object cache.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43782 WIP - experimentation
* #42930 Vulkan command buffer and pool.
* #42709 Vulkan resource cache.
* #42642 Vulkan descriptor and descriptor layout cache.
* #42395 Vulkan pipeline and pipeline layout cache.
* #42786 Vulkan memory allocator.
* #42325 Vulkan (source and binary) shader and shader layout cache.
* **#42394 Generic Vulkan object cache.**
* #42503 Move torch/csrc/utils/hash.h to c10/util/hash.h.
* #43628 Revert "Revert D23252335: Refactor Vulkan context into its own files. Use RAII."

Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Cache.h:
```diff
@@ -0,0 +1,84 @@
+class Cache final {
+ public:
+  explicit Cache(Factory factory);
+  ~Cache() = default;
```
Cache owns map that owns Handles with non trivial destructors(Deleters).
Should it have deleted copy-ctor and copy-assignment to prevent copy and double deletion of Handle's?

AshkanAliabadi replied:
Good point, and even better thinking!  I agree that defining the copy constructor and assignment operator as deleted would have been better practice and that's what I will do.

Having said that, there is no danger of double deletion.  That is because implicitly declared copy constructors (copy assignment operators) are defined as deleted if, among other conditions, at least one non-static data member of the class is not copyable (copy assignable).

From https://en.cppreference.com/w/cpp/language/copy_constructor:

> The implicitly-declared or defaulted copy constructor for class T is  defined as deleted if any of the following conditions are true:
> 
> -  T has non-static data members that cannot be copied (have deleted, inaccessible, or ambiguous copy constructors);
> -  T has direct or virtual base class that cannot be copied (has deleted, inaccessible, or ambiguous copy constructors);
> -  T has direct or virtual base class with a deleted or inaccessible destructor; 

That means all it takes for this class to end up having its implicitly declared copy constructor (copy assignment operator) marked as deleted is for either [one of its member variables] flat_hash_map or Factory to not be copyable (copy assignable), and flat_hash_map will in turn be non-copyable (non-copy assignable) if the data type it contains is not copyable (not copy-assignable), and so on.  Since at::native::vulkan::api::Handle has both copy constructor and assignment deleted, this whole chain will end up being non-copyable.

But I agree that making this explicit is better practice.  I will change the code accordingly.

Commit:
Update on "Generic Vulkan object cache."




[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."




[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]

Commit:
Update on "Generic Vulkan object cache."


Differential Revision: [D23252340](https://our.internmc.facebook.com/intern/diff/D23252340)

[ghstack-poisoned]
